### PR TITLE
fix: render the procedure definition block like Scratch

### LINKED ISSUE #93: procedure definition blocks are not rendered on the top and right edges
The procedure definition blocks are not rendered on the top and right edges as per Scratch.

https://screenshot.googleplex.com/6bVT8eDe94yHKjw

Scratch:

Image

### PR DESCRIPTION
**Pull request by gonfunko**
This PR renders procedure (My Blocks) definition blocks in the Scratch style, with rounded corners and straight edges. This fixes #93.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -65,6 +66,8 @@ export { CheckableContinuousFlyout };
 export { ScratchVariables };
 
 export function inject(container, options) {
+  options["renderer"] = "scratch";
```
Comment (BeksOmega) [inline src/index.js]: Can this go inside of the `Object.assign` call?

File: src/renderer/bowler_hat.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as Blockly from "blockly/core";
+
+export class BowlerHat extends Blockly.blockRendering.Hat {
+  constructor(constants) {
+    super(constants);
+    // Calculated dynamically by computeBounds_().
+    this.width = undefined;
```
Comment (BeksOmega) [inline src/renderer/bowler_hat.js]: mmm I think this is typed as `number`, can you set it to be `0` by default?

File: src/renderer/drawer.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      super.drawStatementInput_(row);
+    }
+  }
+
+  drawRightSideRow_(row) {
+    if (
+      this.info_.isBowlerHatBlock() &&
+      Blockly.blockRendering.Types.isSpacer(row)
+    ) {
+      Blockly.blockRendering.Drawer.prototype.drawRightSideRow_.call(this, row);
```
Comment (BeksOmega) [inline src/renderer/drawer.js:28]: Assuming the base implementation here just happens to have the behavior you want? What does it actually do?

File: src/renderer/drawer.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      this.positionStatementInputConnection_(row);
+    } else {
+      super.drawStatementInput_(row);
+    }
+  }
+
+  drawRightSideRow_(row) {
+    if (
+      this.info_.isBowlerHatBlock() &&
+      Blockly.blockRendering.Types.isSpacer(row)
```
Comment (BeksOmega) [inline src/renderer/drawer.js:24]: This looks like it probably runs into problems if your bowlerhat block has more than one row. Can you just comment that explicitly if that's the case?

File: src/renderer/drawer.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    }
+  }
+
+  drawTop_() {
+    super.drawTop_();
+    // This is a horrible hack, but the superclass' implementation of drawTop_()
+    // provides no way to cleanly override a hat's path without copying and
+    // pasting the entire implementation here. We know that there will only be
+    // one hat on a block, and its path is a known constant, so we just find and
+    // replace it with the desired bowler hat path here.
```
Comment (BeksOmega) [inline src/renderer/drawer.js:40]: 😢

but correct!

You can also link this to https://github.com/google/blockly/issues/7292 if it makes you feel better.

File: src/renderer/render_info.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    super.computeBounds_();
+    if (this.isBowlerHatBlock()) {
+      // Resize the render info to the same width as the widest part of a
+      // bowler hat block.
+      const statementRow = this.rows.find((r) => r.hasStatement);
+      this.width =
+        statementRow.widthWithConnectedBlocks -
+        statementRow.elements.find((e) =>
+          Blockly.blockRendering.Types.isInput(e)
+        ).width +
```
Comment (BeksOmega) [inline src/renderer/render_info.js:45]: Why are you subtracting the width of the inner statement block's input? `statementRow` is the block that lets you see what your caller will look like right?

File: src/renderer/renderer.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  makeDrawer_(block, info) {
+    return new Drawer(block, info);
+  }
+
+  makeRenderInfo_(block) {
+    return new RenderInfo(this, block);
+  }
+}
+
+Blockly.blockRendering.register("scratch", ScratchRenderer);
```
Comment (BeksOmega) [inline src/renderer/renderer.js:21]: Maybe `scratch_vertical` for future proofing?

File: src/renderer/render_info.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  getSpacerRowHeight_(prev, next) {
+    if (this.isBowlerHatBlock() && prev === this.topRow) {
+      return 0;
+    }
+
+    return super.getSpacerRowHeight_(prev, next);
+  }
+
+  getElemCenterline_(row, elem) {
+    if (this.isBowlerHatBlock() && Blockly.blockRendering.Types.isField(elem)) {
```
Comment (BeksOmega) [inline src/renderer/render_info.js:80]: Just curious what this was fixing.

**Review by BeksOmega — APPROVED** [review]

File: src/renderer/drawer.js (hunk 3 above)
Comment (gonfunko) [reply src/renderer/drawer.js:28]: Yeah, it just draws a vertical line.

File: src/renderer/render_info.js (hunk 8 above)
Comment (gonfunko) [reply src/renderer/render_info.js:80]: The "define" text on the block, and the text on the sample procedure block shown next to it, had different baselines, and this adjusts them to be the same.

File: src/renderer/renderer.js (hunk 7 above)
Comment (gonfunko) [reply src/renderer/renderer.js:21]: I think we're just going to delete the horizontal renderer entirely?

File: src/index.js (hunk 1 above)
Comment (gonfunko) [reply src/index.js]: Done!

File: src/renderer/bowler_hat.js (hunk 2 above)
Comment (gonfunko) [reply src/renderer/bowler_hat.js]: Done

File: src/renderer/drawer.js (hunk 4 above)
Comment (gonfunko) [reply src/renderer/drawer.js:24]: Done!

File: src/renderer/drawer.js (hunk 5 above)
Comment (gonfunko) [reply src/renderer/drawer.js:40]: Noted in the comment, thanks!

File: src/renderer/render_info.js (hunk 6 above)
Comment (gonfunko) [reply src/renderer/render_info.js:45]: The math works out this way 🤷‍♂️ the thing being subtracted is the width of the "define" text, and after spending a day trying to understand it and verifying that this worked reliably across languages/block sizes I gave up.

Comment (BeksOmega) [reply src/renderer/render_info.js:45]: Herm it feels very likely to that this is not the thing we want to be subtracting, and instead the width of the "define" text just happens to match some other constant somewhere that makes this look correct. If you change the "define" text to "supercalifragilisticexpialidocious" does the math still work correctly?

Comment (gonfunko) [reply src/renderer/render_info.js:45]: Yep!
Screenshot 2024-08-08 at 2 30 40 PM

Comment (BeksOmega) [reply src/renderer/render_info.js:45]: sweetness sgtm then!

**Review by BeksOmega — APPROVED** [review]

**Pull request merged**